### DISCUSSION TITLE
for NaN, y1 and y2 with sort is different from y

### DISCUSSION THREAD

**Discussion by aboveyunhai** [observablehq/plot#2172] [Q&A] [answered]
I was wondering why `NaN` or invalid value was treat as maximum for `y1` and `y2`, while it's the minimum in `y`,
I could do something like 
```ts
y2: (d) => isNaN(d.population) ? -Infinity : d.population
```
to fake the minimum or maximum, but feel like abusing the property, or maybe I misunderstood how `y1 | y2` works?
```ts
timeseries = [
  {year: 2014, population: 7295.290765},
  {year: 2015, population: 7379.797139},
  {year: 2016, population: 7464.022049},
  {year: 2017, population: 7547.858925},
  {year: 2018, population: NaN},
  {year: 2019, population: 7713.468100},
  {year: 2020, population: 7794.798739}
];
```
```ts
Plot.barY(timeseries, {
  x: "year", 
  y: "population",
  sort: { x: "y" }
}).plot()
```
![image](https://github.com/user-attachments/assets/194bf7bf-ed27-47b4-a1fb-4cb8975ba00d)

```ts
Plot.barY(timeseries, {
  x: "year", 
  y1: 0,
  y2: (d) => d.population,
  sort: { 
     x: "y2" // "y" 
   }
}).plot()
```
![image](https://github.com/user-attachments/assets/7d8e9c2f-11d6-4166-b5ea-2274cfb039ae)

**Comment by Fil** [comment, marked as answer]
The difference is that if you specify `y` you get implicit stacking, and the invalid datum is stacked as a rectangle of height equal to 0. Whereas if you specify `y2`, there is no stacking, thus y2 is NaN and the invalid datum is discarded. The difference is visible if you set `stroke: "red", inset: -1,`.

(Note that it is not "minimum" nor "maximum": Plot and D3 put invalid values at the end when they sort. For instance if you say `sort: { x: "-y2" }` to sort by descending y2, the invalid datum will still be to the right.)

**Reply by aboveyunhai** [reply]
I see, then what would be the proper solution to address this issue? 
Just a custom function inside the sort to replicate the sorting and make invalid to be in the front?

**Reply by Fil** [reply]
The easiest way is to indicate the domain explicitly in the definition of the scale, like so:

```
Plot.plot({
  x: {domain: [2014, 2015…]}
  ...
})
```

or with a function that orders the values as you see fit: 

```
Plot.plot({
  x: {domain: d3.sort(timeseries, d => isNaN(d), d => d.population).map(d => d.year)}
  ...
})

```
